File: jdroid-android-firebase-fcm/src/main/java/com/jdroid/android/firebase/fcm/FcmListenerResolver.java
package com.jdroid.android.firebase.fcm;

import com.google.firebase.messaging.RemoteMessage;
import com.jdroid.android.application.AbstractApplication;
import com.jdroid.java.utils.LoggerUtils;

import org.slf4j.Logger;

public class FcmListenerResolver {

	private static final Logger LOGGER = LoggerUtils.getLogger(FcmListenerResolver.class);

	/**
	 * Called when message is received. Since Android O, have a guaranteed life cycle limited to 10 seconds for this method execution
	 */
	public void onMessageReceived(RemoteMessage remoteMessage) {

		if (LoggerUtils.isEnabled()) {
			StringBuilder builder = new StringBuilder();
			builder.append("Message received. ");
			builder.append("from: ");
			builder.append(remoteMessage.getFrom());

			if (remoteMessage.getMessageType() != null ) {
				builder.append(", ");
				builder.append("message type: ");
				builder.append(remoteMessage.getMessageType());
			}

			if (remoteMessage.getData() != null & !remoteMessage.getData().isEmpty()) {
				builder.append(", ");
				builder.append("data: ");
				builder.append(remoteMessage.getData());
			}
			LOGGER.info(builder.toString());
		}

		FcmMessageResolver fcmResolver = AbstractFcmAppModule.get().getFcmMessageResolver(remoteMessage.getFrom());
		if (fcmResolver != null) {
			FcmMessage fcmMessage = null;
			try {
				fcmMessage = fcmResolver.resolve(remoteMessage);
			} catch (Exception e) {
				AbstractApplication.get().getExceptionHandler().logHandledException("Error when resolving FCM message", e);
				return;
			}

			if (fcmMessage != null) {
				try {
					fcmMessage.handle(remoteMessage);
				} catch (Exception e) {
					AbstractApplication.get().getExceptionHandler().logHandledException("Error when handling FCM message", e);
				}
			} else {
				LOGGER.warn("The FCM message was not resolved");
			}
		} else {
			LOGGER.warn("A FCM message was received, but not resolved is configured");
		}
	}

	public void onMessageSent(String msgId) {
		LOGGER.info("Message sent with id: " + msgId);
	}

	public void onSendError(String msgId, Exception exception) {
		AbstractApplication.get().getExceptionHandler().logWarningException("Send error. Message id: " + msgId, exception);
	}

	public void onDeletedMessages() {
		LOGGER.info("Deleted messages");
	}
	
	public void onNewToken(String token) {
		LOGGER.info("New token: " + token);
		AbstractFcmAppModule.get().startFcmRegistration(false);
	}
	
}

File: jdroid-android-firebase-fcm/src/main/java/com/jdroid/android/firebase/fcm/FcmListenerResolver.kt
package com.jdroid.android.firebase.fcm

import com.google.firebase.messaging.RemoteMessage
import com.jdroid.android.application.AbstractApplication
import com.jdroid.java.utils.LoggerUtils

class FcmListenerResolver {

    /**
     * Called when message is received. Since Android O, have a guaranteed life cycle limited to 10 seconds for this method execution
     */
    fun onMessageReceived(remoteMessage: RemoteMessage) {

        if (LoggerUtils.isEnabled()) {
            val builder = StringBuilder()
            builder.append("Message received. ")
            builder.append("from: ")
            builder.append(remoteMessage.from)

            if (remoteMessage.messageType != null) {
                builder.append(", ")
                builder.append("message type: ")
                builder.append(remoteMessage.messageType)
            }

            if ((remoteMessage.data != null) and remoteMessage.data.isNotEmpty()) {
                builder.append(", ")
                builder.append("data: ")
                builder.append(remoteMessage.data)
            }
            LOGGER.info(builder.toString())
        }

        val fcmResolver = AbstractFcmAppModule.get().fcmMessageResolver
        var fcmMessage: FcmMessage?
        try {
            fcmMessage = fcmResolver.resolve(remoteMessage)
        } catch (e: Exception) {
            AbstractApplication.get().exceptionHandler.logHandledException("Error when resolving FCM message", e)
            return
        }

        if (fcmMessage != null) {
            try {
                fcmMessage.handle(remoteMessage)
            } catch (e: Exception) {
                AbstractApplication.get().exceptionHandler.logHandledException("Error when handling FCM message", e)
            }
        } else {
            LOGGER.warn("The FCM message was not resolved")
        }

        for (fcmEventsListener in FcmContext.getFcmEventsListeners()) {
            try {
                fcmEventsListener.onMessageReceived(remoteMessage)
            } catch (e: Exception) {
                AbstractApplication.get().exceptionHandler.logHandledException(e)
            }
        }
    }

    fun onMessageSent(msgId: String) {
        LOGGER.info("Message sent with id: $msgId")

        for (fcmEventsListener in FcmContext.getFcmEventsListeners()) {
            try {
                fcmEventsListener.onMessageSent(msgId)
            } catch (e: Exception) {
                AbstractApplication.get().exceptionHandler.logHandledException(e)
            }
        }
    }

    fun onSendError(msgId: String, exception: Exception) {
        AbstractApplication.get().exceptionHandler.logWarningException("Send error. Message id: $msgId", exception)

        for (fcmEventsListener in FcmContext.getFcmEventsListeners()) {
            try {
                fcmEventsListener.onSendError(msgId, exception)
            } catch (e: Exception) {
                AbstractApplication.get().exceptionHandler.logHandledException(e)
            }
        }
    }

    fun onDeletedMessages() {
        LOGGER.info("Deleted messages")

        for (fcmEventsListener in FcmContext.getFcmEventsListeners()) {
            try {
                fcmEventsListener.onDeletedMessages()
            } catch (e: Exception) {
                AbstractApplication.get().exceptionHandler.logHandledException(e)
            }
        }
    }

    fun onNewToken(token: String) {
        LOGGER.info("New token: $token")

        for (fcmEventsListener in FcmContext.getFcmEventsListeners()) {
            try {
                fcmEventsListener.onNewToken(token)
            } catch (e: Exception) {
                AbstractApplication.get().exceptionHandler.logHandledException(e)
            }
        }

        AbstractFcmAppModule.get().startFcmRegistration(false)
    }

    companion object {
        private val LOGGER = LoggerUtils.getLogger(FcmListenerResolver::class.java)
    }
}
